src/Connection/Client.php

```diff
@@ -44,7 +65,105 @@ public function syncDisconnect(): bool
 		}
 
 		$this->init();
-		
+
 		return true;
 	}
+
+	protected function write(): void
+	{
+		parent::write();
+		if (($last = error_get_last()) !== null) {
+			if (!Strings::match($last['message'], '~fwrite(): Send of \d+ bytes failed with errno=\d+ broken pipe~i')) {
+				return;
+			}
+
+			error_clear_last();
+			throw new ClientException('Broken pipe or closed connection.');
+		}
+	}
+
+	/**
+	 * Runs it's own event loop, processes frames as they arrive. Processes messages for at most $maxSeconds.
+	 *
+	 * @param float $maxSeconds
+	 */
+	public function run($maxSeconds = null): void
+	{
+		if (!$this->isConnected()) {
+			throw new ClientException('Client has to be connected.');
+		}
+
+		$this->running = true;
+		$startTime = microtime(true);
+		$stopTime = null;
+		if ($maxSeconds !== null) {
+			$stopTime = $startTime + $maxSeconds;
+		}
+
+		do {
+			if (is_callable($this->options['cycle_callback'] ?? null)) {
```

Comment:
There are multiple is_callable checks for basically 'readonly' thing... I would move that check to constructor, throw Exception in case 'cycle_callback' is not callable

Reply:
And when is_callable check is gone next line could be 
`$this->options['cycle_callback']?->call($this);`